url49584c1d41_Arrem_GGS_tgz-pJ8-com_gamezgalaxy_test_console_MainJ8/src/com/gamezgalaxy/GGS/world/convert/DatToGGS.java

```diff
@@ -0,0 +1,181 @@
+		try {
+			fos = new FileOutputStream(filename);
+			gzos = new GZIPOutputStream(fos);
+			outputstream = new DataOutputStream(gzos);
+			outputstream.writeInt(0x271bb788);
+			outputstream.writeByte(2);
+			out = new ObjectOutputStream(gzos);
+			out.writeObject(level);
+			outputstream.close();
+			out.close();
+			System.out.println("Saving level "+filename+" successful");
+		} catch(IOException ex) {
+			ex.printStackTrace();
+		}
+	}
+
+	// prints all there is to know about a level, except for the blocks data
+	public void printInfo() {
+		if (level == null) {
+			return;
+		}
+		System.out.println("Level info:");
+		System.out.println("name: "+level.name);
+		System.out.println("creator: "+level.creator);
+		System.out.println("createTime: "+(DateFormat.getDateTimeInstance(DateFormat.FULL,DateFormat.FULL).format(level.createTime)));
+		System.out.println("width: "+level.width);
+		System.out.println("height: "+level.height);
+		System.out.println("depth: "+level.depth);
+		System.out.println("spawnpoint: ["+level.xSpawn+","+level.ySpawn+","+level.zSpawn+"]");
+		System.out.println("spawn rotation: "+level.rotSpawn);
+	}
+
+	// safe to use method, return value let's you know if anything was changed
+	public boolean setTile(int x, int y, int z, int t) {
+		if (
+			x >=0 && x < level.width &&
+			y >=0 && y < level.depth &&
+			z >=0 && z < level.height &&
+			t >= 0 && t <= 37
+		) {
+			if (t == 8 || t == 10) {
+				level.setTile(x,y,z,t);
+			} else if (t >= 0 && t <= 18) {
+				level.setTileNoUpdate(x,y,z,t);
+			}
+			return true;
+		}
+		return false;
+	}
+
+	// gets you the level coordinates from the blocks array index
+	public int[] getCoords(int index) {
+		int x = index % width;
+		index = (index-x) / width;
+		int z = index % height;
+		int y = (index-z) / height;
+		return new int[] {x, y, z};
+	}
+
+	public void clearBlocks() {
+		for (int i=0; i<level.blocks.length; i++) {
+			level.blocks[i] = 0;
+		}
+	}
+
+	public void floor(int y, int type) {
+		for (int i=0; i<level.width; i++) {
+		for (int j=0; j<level.height; j++) {
+			setTile(i,y,j,type);
+		}
+		}
+	}
+
+	public void wallX(int x1, int x2, int z, int y, int height, int type) {
+		for (int i=x1; i<=x2; i++) {
+		for (int j=y; j<y+height; j++) {
+			if (!setTile(i,j,z,type)) {
+				System.out.println("Warning: a tile got ignored while building a wallX: ["+i+","+j+","+z+"]");
+			}
+		}
+		}
+	}
+
+	public void wallZ(int x, int z1, int z2, int y, int height, int type) {
+		for (int i=z1; i<=z2; i++) {
+		for (int j=y; j<y+height; j++) {
+			if (!setTile(x,j,i,type)) {
+				System.out.println("Warning: a tile got ignored while building a wallZ: ["+x+","+j+","+i+"]");
+			}
+		}
+		}
+	}
+
+	// substitute all of block type 'from' to 'to' :) returning the number of blocks altered
+	public int substitute(byte from, byte to) {
+		int count=0;
+		for (int i=0; i<level.blocks.length; i++) {
+			if (level.blocks[i] == from) {
+				level.blocks[i] = to;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;to&#39; (index: 1 | type: byte). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other validated parameters: &#39;from&#39;. The same parameter type is validated here for example: url49584c1d41_Arrem_GGS_tgz-pJ8-com_gamezgalaxy_test_console_MainJ8/src/com/gamezgalaxy/GGS/networking/packets/browser/GET.java:22. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, [reduce the method&#39;s visibility](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html).

 Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).